src/Emulator/Cores/Arm-M/NVIC.cs

```diff
@@ -539,15 +548,18 @@ private int FindPendingInterrupt()
                         this.NoisyLog("IRQ {0} preempts {1}.", result, activeIRQs.Peek());
                     }
                 }
-                IRQ.Set(!PRIMASK && result != SpuriousInterrupt);
```

Comment:
I came at this thinking that only AcknowledgeIRQ() would be reponsible for eventually unsetting the IRQ, and that FindPendingInterrupt may only ever be in situations where it'd set  IRQ from false->true or true->true, if that makes sense?

Reply:
I am not yet sure why does the test software not go the usual way. I will look more into this, as it is quite surprising.

Reply:
just curious if you have been able to dig into this one more? BTW I think I lost track of what are the different changes you think should be done, I've taken note of the change from HashSet to SortedSet, but not 100% sure what the other ones should be :)

Reply:
Yes, I did spend some time on this. Unfortunately it's not super trivial to pinpoint, as the fail in my test case does not happen immediately.

My best bet at the moment is that this fail is related to BASEPRI being set. I am not 100% sure yet, but it does look sensible.

Honestly speaking I'm inclined towards preserving the previous logic instead of digging deeper here. I don't think it should influence your performance very much.

Reply:
Would it be possible to share the tests that are failing, by any chance? The impact on performance is about 2x so it is still something that I hope can be fixed eventually (that is, unless the logic is not right in the first place, and IRQ should indeed be forcefully unset as part of FindPendingInterrupt)

Reply:
Oh, that's very significant! I wasn't aware this change had such an impact.

Well, there are three possible cases here:
1) the test is wrong
2) the algorithm in NVIC is wrong in other places
3) your speed-up is due to interrupts appearing not often enough.

I think we can rule out option 1 with, let's say, 80% confidence. We use a Zephyr-based test, it's not really fancy.
The test is here: https://github.com/renode/renode/blob/master/tests/platforms/NRF52840.robot#L88
I'll try to dig up the sources for this test, but's it's a generic Zephyr sample with some configuration.

Option 2 is possible of course.

Option 3: I'm not sure. The application might not even notice that some interrupts are appearing less often.

Reply:
You can see it run by `./test.sh --show-log -f "*NRF52840.Should Run ZephyrRTOS Shell On UART*" tests/platforms/NRF52840.robot` (unfortunately it requires ctrl+c after running)